### PR TITLE
Added test fixture class to get transport stats

### PR DESCRIPTION
**Pull request by AbhayPatil**

### CODE REVIEW

File: protocol_driver_test_fixture.cc [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+void ProtocolDriverTestFixture::HandleConnectFailure(
+    std::string_view local_connection_info) {
+  pd_instance_->HandleConnectFailure(local_connection_info);
+}
+
+std::vector<TransportStat> ProtocolDriverTestFixture::GetTransportStats() {
+  std::vector<TransportStat> transport_stats;
+  transport_stats.push_back({"client_rpc_cnt", client_rpc_cnt});
```
Comment (danmanj) [inline protocol_driver_test_fixture.cc:96]: you should be calling the method of the pd below you as well.

File: protocol_driver_test_fixture.cc [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+absl::Status ProtocolDriverTestFixture::InitializeServer(
+    const ProtocolDriverOptions& pd_opts, int* port) {
+  return absl::UnimplementedError("InitializeServer is not implemented for test_fixture.");
+}
+
+void ProtocolDriverTestFixture::SetHandler(
+    std::function<std::function<void()>(ServerRpcState* state)> handler) {
+  auto server_handler = [=] (ServerRpcState* state) {
+    ++server_rpc_cnt;
```
Comment (danmanj) [inline protocol_driver_test_fixture.cc:66]: explicit memory order

File: protocol_driver_test_fixture.cc [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+absl::Status ProtocolDriverTestFixture::InitializeClient(
+    const ProtocolDriverOptions& pd_opts) {
+  return absl::UnimplementedError("InitializeClient is not implemented for test_fixture.");
+}
+
+absl::Status ProtocolDriverTestFixture::InitializeServer(
+    const ProtocolDriverOptions& pd_opts, int* port) {
+  return absl::UnimplementedError("InitializeServer is not implemented for test_fixture.");
```
Comment (danmanj) [inline protocol_driver_test_fixture.cc:60]: Just call the method.

File: protocol_driver_test_fixture.cc [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+  auto maybe_pd_instance_ = AllocateProtocolDriver(pdo, &port_, tree_depth_+1);
+  if (!maybe_pd_instance_.ok()) return maybe_pd_instance_.status();
+  pd_instance_ = std::move(maybe_pd_instance_.value());
+
+  return absl::OkStatus();
+}
+
+absl::Status ProtocolDriverTestFixture::InitializeClient(
+    const ProtocolDriverOptions& pd_opts) {
+  return absl::UnimplementedError("InitializeClient is not implemented for test_fixture.");
```
Comment (danmanj) [inline protocol_driver_test_fixture.cc:55]: call the method.

File: protocol_driver_test_fixture.cc [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+  tree_depth_ = tree_depth;
+}
+
+absl::Status ProtocolDriverTestFixture::Initialize(
+    const ProtocolDriverOptions& pd_opts, int* port) {
+
+  auto pdo = pd_opts;
+  auto server_settings = pdo.mutable_server_settings();
+  google::protobuf::RepeatedPtrField<NamedSetting>::iterator
+      next_protocol_driver_it = server_settings->end();
```
Comment (danmanj) [inline protocol_driver_test_fixture.cc:36]: name doesnt match things its looking for

File: protocol_driver_test_fixture.cc [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+  for (auto it=server_settings->begin(); it!=server_settings->end(); it++) {
+      if (it->name()=="driver_under_test") {
+        next_protocol_driver_it = it;
+        pdo.set_protocol_name(it->string_value());
+      }
+  }
+  if (next_protocol_driver_it != server_settings->end())
+    server_settings->erase(next_protocol_driver_it);
+
+  auto maybe_pd_instance_ = AllocateProtocolDriver(pdo, &port_, tree_depth_+1);
```
Comment (danmanj) [inline protocol_driver_test_fixture.cc:46]: don't declare port_ you can just use the port pointer that is passed in.

File: protocol_driver_test.cc [hunk 7]
```diff
@@ -114,6 +114,20 @@ TEST_P(ProtocolDriverTest, Invoke) {
     });
   }
   pd->ShutdownClient();
+
+  if (pdo.protocol_name() == "double_barrel") {
```
Comment (danmanj) [inline protocol_driver_test.cc:118]: this should be a separate test function.

File: protocol_driver_test_fixture.h [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+                   std::function<void(void)> done_callback) override;
+  void ChurnConnection(int peer) override;
+  void ShutdownServer() override;
+  void ShutdownClient() override;
+
+ private:
+  std::unique_ptr<distbench::ProtocolDriver> pd_instance_;
+  int port_ = 0;
+  int tree_depth_;
+  std::atomic<int> client_rpc_cnt{0};
```
[the remark this answers is not in the document]
Comment (AbhayPatil) [reply protocol_driver_test_fixture.h:57]: Done

File: protocol_driver_test_fixture.cc [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+  std::vector<TransportStat> transport_stats;
+  transport_stats.push_back({"client_rpc_cnt", client_rpc_cnt});
+  transport_stats.push_back({"server_rpc_cnt", server_rpc_cnt});
+  return transport_stats;
+}
+
+void ProtocolDriverTestFixture::InitiateRpc(int peer_index, ClientRpcState* state,
+                                     std::function<void(void)> done_callback) {
+  pd_instance_->InitiateRpc(peer_index, state, done_callback);
+  ++client_rpc_cnt;
```
[the remark this answers is not in the document]
Comment (AbhayPatil) [reply protocol_driver_test_fixture.cc:104]: Done

File: protocol_driver_test_fixture.cc (hunk 2 above)
Comment (AbhayPatil) [reply protocol_driver_test_fixture.cc:66]: Done

File: protocol_driver_test_fixture.cc (hunk 6 above)
Comment (AbhayPatil) [reply protocol_driver_test_fixture.cc:46]: Done

File: protocol_driver_test_fixture.cc [hunk 10]
```diff
@@ -0,0 +1,119 @@
+// Copyright 2021 Google LLC
```
Comment (oserres) [inline protocol_driver_test_fixture.cc:1]: Should be 2022

File: protocol_driver_test_fixture.h [hunk 11]
```diff
@@ -0,0 +1,63 @@
+// Copyright 2021 Google LLC
```
Comment (oserres) [inline protocol_driver_test_fixture.h:1]: 2022

**Review by oserres — APPROVED** [review]
Code is badly formatted; also run
make clang-format

File: protocol_driver_test_fixture.cc (hunk 1 above)
Comment (AbhayPatil) [reply protocol_driver_test_fixture.cc:96]: Done

File: protocol_driver_test_fixture.cc (hunk 10 above)
Comment (AbhayPatil) [reply protocol_driver_test_fixture.cc:1]: Done

Comment (AbhayPatil) [reply protocol_driver_test_fixture.cc:1]: Have to run clang-format on lot of files. Let us do that outside current pull request ?

File: protocol_driver_test.cc (hunk 7 above)
Comment (AbhayPatil) [reply protocol_driver_test.cc]: Made a new file composable_protocol_driver_test.cc

File: composable_rpc_counter.cc [hunk 12]
```diff
@@ -43,76 +43,78 @@ absl::Status ProtocolDriverTestFixture::Initialize(
...
   pd_instance_ = std::move(maybe_pd_instance_.value());
 
   return absl::OkStatus();
 }
 
-absl::Status ProtocolDriverTestFixture::InitializeClient(
+absl::Status ComposableRpcCounter::InitializeClient(
     const ProtocolDriverOptions& pd_opts) {
-  return absl::UnimplementedError("InitializeClient is not implemented for test_fixture.");
+  return absl::UnimplementedError("InitializeClient should not be implemented here because Initialize() is invoked directly from AllocateProtocolDriver().");
```
Comment (danmanj) [inline composable_rpc_counter.cc:55]: the long message could just be a comment, not a dissertation.

**Review by danmanj — APPROVED** [review]

**Pull request merged**